### PR TITLE
[flash_ctrl] ADD TEST FLASH CTRL ERROR PROG WIN

### PR DESCRIPTION
**Pull request by TimEwins**
Perform accesses in order to provoke the 'program resolution' error.
Test via the Software interface.

Signed-off-by: TIM EWINS <tim.ewins@ensilica.com>

### CODE REVIEW

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv]: Fixed

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:32]: Changed

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_base_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_base_vseq.sv]: Fixed

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (eshapira) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:86]: As in the other comment, this shouldn't be dependent of flash_op operation, since it's always a program:
```suggestion
    solve flash_op before flash_op_data;
    flash_op_data.size() == flash_op.num_words;
```

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (eshapira) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:188]: If you do only transactions to the data partition, no need for those.
```suggestion
```

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (eshapira) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:70]: In this test you're only using program transactions and only to data partition. correct?
If so, no need for all of those constraints.
```suggestion
```

**Review by eshapira — APPROVED** [review]
LGTM except one thing.
Please note that this thing is relevant also for your other PR #12310.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:86]: Hi @eshapira I agree with everything you say, and will change the code accordingly.

The method of having a common set of random constraints and then using 'randomize with' is another way of doing it.

There is already an issue raised on us to look at putting a default set of constraints into the base class, as there is a ot of repetition of constraints as you have observed.

I have a feeling that this will eventually change again.

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:188]: As above

**Comment by TimEwins** [comment]
Please do not merge until I push new changes

**Comment by TimEwins** [comment]
I have updated the following ....

1) Simplified the constraints, as suggested
2) Removed some $urandom expressions and replaced them with class constraints for better seed stability

Ready for Review

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:41]: Done

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:43]: Done

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:217]: Done

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (jdonjdon) [inline hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:1]: Can you add this file to https://github.com/lowRISC/opentitan/pull/12310/files ?
Unless someone ask you to create pr per vseq, I don't see why we have 2 PRs' with mostly redundant
updates.

**Review by jdonjdon — CHANGES_REQUESTED** [review]

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:1]: The PR submission 'process' we are using 'limits' the number of code lines that can be submitted in one PR.  Having one PR per vseq compartmenatlises the submissions which are sequential in this instance.

So it is best to stick to two PRs, especially now they are both already submitted.

**Comment by TimEwins** [comment]
Resubmitted

**Comment by TimEwins** [comment]
Some checks are failing ... seems a common proble.  Suspect issues outside this PR.

**Comment by TimEwins** [comment]
Please can someone retrigger the checkin checks to see if the errors have cleared ?

File: hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv
Comment (weicaiyang) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:1]: One vseq one PR is probably fine. And we actually encourage people to make small PRs. the same changes on base_vseq will be merged after both PR are landed.

Comment (TimEwins) [reply hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_error_prog_win_vseq.sv:1]: Sometimes there is a need to duplicate in order to make progress.  The PR review system usually takes a while, and we need to make forward progress.

**Comment by TimEwins** [comment]
Rebased and ready for merge

This merge will conflict with #12310

Depending on which is merged first the other will need rebasing

**Comment by TimEwins** [comment]
Checks need to be re-ran .. something is failing elsewhere

**Comment by weicaiyang** [comment]
there is a conflicting file. @TimEwins can you resolve it?

**Review by jdonjdon — APPROVED** [review]

**Comment by TimEwins** [comment]
@weicaiyang Rebased and clean ... subject to further PR merges if they occur in the meantime

**Comment by weicaiyang** [comment]
> @weicaiyang Rebased and clean ... subject to further PR merges if they occur in the meantime

yeah, need to rebase again...

**Review by weicaiyang — APPROVED** [review]

**Comment by TimEwins** [comment]
@weicaiyang I'll do the rebase now

**Comment by TimEwins** [comment]
@weicaiyang Ready for merge

**Pull request merged**
